File: lafza/backend/app/services/parent_summary.py
"""Aggregations behind the parent dashboard and weekly report (P1/P3).

Streak counts consecutive calendar days (UTC) with at least one scored
utterance; practice time sums session durations over the last 7 days.
"""

import uuid
from datetime import datetime, timedelta, timezone

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import PhonemeProfile, TherapySession, TreatmentPlan

WEEK_DAYS = 7


def build_parent_summary(db: Session, child_id: uuid.UUID) -> dict:
    now = datetime.now(timezone.utc)
    today = now.date()
    week_start = now - timedelta(days=WEEK_DAYS - 1)

    profiles = list(
        db.scalars(
            select(PhonemeProfile).where(PhonemeProfile.child_id == child_id)
        )
    )

    # Streak: consecutive active days ending today (or yesterday, so an
    # unfinished today doesn't break the chain).
    active_days = {p.created_at.date() for p in profiles}
    cursor = today if today in active_days else today - timedelta(days=1)
    streak = 0
    while cursor in active_days:
        streak += 1
        cursor -= timedelta(days=1)

    sessions = list(
        db.scalars(
            select(TherapySession).where(
                TherapySession.child_id == child_id,
                TherapySession.created_at >= week_start,
            )
        )
    )
    seconds_by_day: dict = {}
    for session in sessions:
        day = session.created_at.date()
        seconds_by_day[day] = (
            seconds_by_day.get(day, 0) + (session.duration_sec or 0)
        )
    daily = [
        {"day": day, "seconds": seconds_by_day.get(day, 0)}
        for i in range(WEEK_DAYS)
        for day in [today - timedelta(days=WEEK_DAYS - 1 - i)]
    ]

    week_profiles = [
        p for p in profiles if p.created_at >= week_start
    ]
    by_phoneme: dict[str, list[PhonemeProfile]] = {}
    for p in sorted(week_profiles, key=lambda r: r.created_at):
        by_phoneme.setdefault(p.phoneme, []).append(p)
    phoneme_rows = sorted(
        (
            {
                "phoneme": phoneme,
                "attempts": len(rows),
                "first_gop": rows[0].gop_score,
                "latest_gop": rows[-1].gop_score,
                "latest_error_type": rows[-1].error_type,
            }
            for phoneme, rows in by_phoneme.items()
        ),
        key=lambda r: r["latest_gop"],  # weakest first
    )

    latest_plan = db.scalars(
        select(TreatmentPlan)
        .where(TreatmentPlan.child_id == child_id)
        .order_by(TreatmentPlan.created_at.desc())
        .limit(1)
    ).first()

    return {
        "streak_days": streak,
        "week_practice_seconds": sum(d["seconds"] for d in daily),
        "week_attempts": len(week_profiles),
        "daily": daily,
        "phonemes": phoneme_rows,
        "plan": (
            {
                "target_phonemes": latest_plan.target_phonemes,
                "status": latest_plan.status,
            }
            if latest_plan
            else None
        ),
    }

File: lafza/mobile/lib/data/models/parent_summary.dart
/// Payload of GET /children/{id}/parent-summary (P1 dashboard + P3 report).
class ParentSummary {
  const ParentSummary({
    required this.streakDays,
    required this.weekPracticeSeconds,
    required this.weekAttempts,
    required this.daily,
    required this.phonemes,
    required this.plan,
  });

  factory ParentSummary.fromJson(Map<String, dynamic> json) => ParentSummary(
        streakDays: json['streak_days'] as int,
        weekPracticeSeconds: json['week_practice_seconds'] as int,
        weekAttempts: json['week_attempts'] as int,
        daily: [
          for (final d in json['daily'] as List)
            DailyPractice.fromJson(d as Map<String, dynamic>),
        ],
        phonemes: [
          for (final p in json['phonemes'] as List)
            PhonemeReportRow.fromJson(p as Map<String, dynamic>),
        ],
        plan: json['plan'] == null
            ? null
            : PlanSummary.fromJson(json['plan'] as Map<String, dynamic>),
      );

  final int streakDays;
  final int weekPracticeSeconds;
  final int weekAttempts;
  final List<DailyPractice> daily; // 7 entries, oldest → today
  final List<PhonemeReportRow> phonemes; // weakest first
  final PlanSummary? plan;
}

class DailyPractice {
  const DailyPractice({required this.day, required this.seconds});

  factory DailyPractice.fromJson(Map<String, dynamic> json) => DailyPractice(
        day: DateTime.parse(json['day'] as String),
        seconds: json['seconds'] as int,
      );

  final DateTime day;
  final int seconds;
}

class PhonemeReportRow {
  const PhonemeReportRow({
    required this.phoneme,
    required this.attempts,
    required this.firstGop,
    required this.latestGop,
    required this.latestErrorType,
  });

  factory PhonemeReportRow.fromJson(Map<String, dynamic> json) =>
      PhonemeReportRow(
        phoneme: json['phoneme'] as String,
        attempts: json['attempts'] as int,
        firstGop: json['first_gop'] as int,
        latestGop: json['latest_gop'] as int,
        latestErrorType: json['latest_error_type'] as String?,
      );

  final String phoneme;
  final int attempts;
  final int firstGop;
  final int latestGop;
  final String? latestErrorType;

  int get delta => latestGop - firstGop;
}

class PlanSummary {
  const PlanSummary({required this.targetPhonemes, required this.status});

  factory PlanSummary.fromJson(Map<String, dynamic> json) => PlanSummary(
        targetPhonemes: [
          for (final p in json['target_phonemes'] as List) p as String,
        ],
        status: json['status'] as String,
      );

  final List<String> targetPhonemes;
  final String status;
}
